### ISSUE TITLE
All your modules will compile!

### ISSUE THREAD

**Issue by dellaert** [borglab/wrap#137]
![image](https://user-images.githubusercontent.com/10515273/145906796-07acb8c6-7de8-40f7-877e-77f2be9ed22a.png)

If I change *one* .i file, *all* modules are re-generated and compiled. Takes *ages* to do substantial work on wrapper that way. Might be a simple cmake edit?

**Comment by dellaert** [comment]
Conversely, if I change a  specialization file (and only that), specialization file in build does not get re-generated. Maybe that's too much to ask though :-)

**Comment by varunagrawal** [comment]
This is interesting. I always thought only a specific module was built (at least in the case of GTSAM).

You may want to turn off the Matlab wrapper since module level optimization is not enabled for it. Only the python wrapper supports per module compilation.

**Comment by dellaert** [comment]
Matlab is off…

**Comment by dellaert** [comment]
@varunagrawal and @ProfFan could you slack about this and see whether there is a fix? Seems re-generation of a module should be dependent only on its .i file, not *any* .i file. This is really holding me back in wrapping discrete right now.

**Comment by varunagrawal** [comment]
Alright, I'll focus on this today and get it resolved.

**Comment by dellaert** [comment]
> Alright, I'll focus on this today and get it resolved.

But chat w Fan first!

**Comment by varunagrawal** [comment]
@dellaert spoke to Fan and I took a look at the reason for this.

We give the wrapper a list of all the .i files and it generates the corresponding .cpp files each time, hence `make` assumes it needs to build all the files each time. We could have cmake generate each .cpp file individually but that is complex and error-prone since very few people have the required CMake expertise compared to Python expertise (which currently generates all the files).

Moreover, cmake doesn't figure out which files have changed, that is `make`'s responsibility, so this is a particularly challenging issue.

One thing I can recommend for now is to go to `python/CMakeLists.txt` and comment out all the interface headers you don't need (starting at line 56). That should significantly improve compilation for you.

**Comment by dellaert** [comment]
Can you just loop of the interface files and generate the dependency that way?

**Comment by dellaert** [comment]
PS @jlblancoc is a cmake expert. Maybe he has an easy fix. It seems that this is exactly what cmake is supposed to do.

**Comment by varunagrawal** [comment]
That's what I was considering. I need to look more into this and would really appreciate @jlblancoc's help. :slightly_smiling_face:

**Comment by jlblancoc** [comment]
I'll take a look at it this weekend!

**Comment by jlblancoc** [comment]
Wow, it's the first time I took a serious look at all the work behind wrap and it's amazing :100: 

Now I see your issue. You should probably do this using several  [`add_custom_target()`](https://cmake.org/cmake/help/latest/command/add_custom_target.html) commands, one per `.i` -> `.cpp` file pair, specifying the `.cpp` as `BYPRODUCTS` and the `.i` as `SOURCES` and `DEPENDS`. 

Next, a final `add_custom_target()` that takes all `.cpp`s as `SOURCES` and does the actual python module compilation.
This latter stage is, I think, the one you already have in PybindWrap.cmake (line 77): 

```cmake
  add_custom_target(pybind_wrap_${module_name} ALL DEPENDS ${cpp_files})
```

BTW: If this `DEPENDS ${cpp_files}` causes problems (I don't think so...), you could also add manual `add_dependencies()` for the `pybind_wrap_${module_name}` target on all the former `.i -> .cpp` targets, to make sure the .cpp are up-to-date.

In short: split your `add_custom_command(... ${PYTHON_EXECUTABLE} ${PYBIND_WRAP_SCRIPT} ... )` into many calls, and let's see if `BYPRODUCTS` handle the dependencies as `make` would do...

Let me know if this makes sense to you... If you draft a test version ping me if needs a hand with it.

**Comment by varunagrawal** [comment]
Yes that makes a lot of sense! Great idea @jlblancoc

**Comment by varunagrawal** [comment]
Okay so I did the updates as recommended by @jlblancoc and I seem to have done something wrong since the wrapper now compiles every time. Any help would be appreciated since I can't seem to find what the issue is. :confused:
